kernel/sched.c

```diff
@@ -454,84 +457,72 @@ void z_sched_start(struct k_thread *thread)
 	z_reschedule(&_sched_spinlock, key);
 }
 
-/**
- * @brief Halt a thread
- *
- * If the target thread is running on another CPU, flag it as needing to
- * abort and send an IPI (if supported) to force a schedule point and wait
- * until the target thread is switched out (ISRs will spin to wait and threads
- * will block to wait). If the target thread is not running on another CPU,
- * then it is safe to act immediately.
- *
- * Upon entry to this routine, the scheduler lock is already held. It is
- * released before this routine returns.
- *
- * @param thread Thread to suspend or abort
- * @param key Current key for _sched_spinlock
- * @param terminate True if aborting thread, false if suspending thread
+/* Spins in ISR context, waiting for a thread known to be running on
+ * another CPU to catch the IPI we sent and halt.  Note that we check
+ * for ourselves being asynchronously halted first to prevent simple
+ * deadlocks (but not complex ones involving cycles of 3+ threads!).
+ * Acts to release the provided lock before returning.
  */
-static void z_thread_halt(struct k_thread *thread, k_spinlock_key_t key,
-			  bool terminate)
+static void thread_halt_spin(struct k_thread *thread, k_spinlock_key_t key)
 {
-#ifdef CONFIG_SMP
-	if (is_halting(_current) && arch_is_in_isr()) {
-		/* Another CPU (in an ISR) or thread is waiting for the
-		 * current thread to halt. Halt it now to help avoid a
-		 * potential deadlock.
-		 */
+	if (is_halting(_current)) {
 		halt_thread(_current,
-			    is_aborting(_current) ? _THREAD_DEAD
-						  : _THREAD_SUSPENDED);
+			    is_aborting(_current) ? _THREAD_DEAD : _THREAD_SUSPENDED);
 	}
+	k_spin_unlock(&_sched_spinlock, key);
+	while (is_halting(thread)) {
+		arch_spin_relax();
+	}
+}
```

Comment:
According to the weak definition of arch_spin_relax() in idle.c, this is intended to be called while interrupts are locked (I've been bitten by this before).

Reply:
Oh, right.  @npitre needed that IIRC because the relax on IIRC arm64 needs to poll for an IPI because of the somewhat frustrating FPU shootdown deadlock (not unlike this one).  This may need to be a hand-coded relaxation.  I realize I'm not hitting it in CI because your patch with the test got dropped a while back, should just add that to this PR.

Reply:
It may look a little clunky, but we could do ..

```
k_spin_release(&_sched_spinlock);
while (is_halting(thread)) {
    arch_spin_relax();

    arch_irq_unlock(key.key);  /* Unlock/lock interrupts for better */
    key.key = arch_irq_lock();  /* nested interrupt responsiveness */
}
arch_irq_unlock(key.key);
```

Reply:
Yeah, that's likely the best way to do it, given that we really do want architectures to be able to define relaxation code on their own (we actually don't need the arm64 magic at all here though, since the released interrupts would allow the FPU IPI to be handled).